lib/src/lib.rs

```diff
@@ -1135,76 +1137,74 @@ impl Repo {
+                        {
+                            let mut sh = shared.lock().unwrap();
 
+                            if sh.in_progress.contains(&path) {
+                                continue;
                             } else {
-                                break;
+                                sh.in_progress.insert(path.clone());
                             }
                         }
-                    }
-                            });
-            }
-            drop(pool_rx);
 
-            loop {
-                match rx.recv() {
-                    Err(_) => {
+                        // check if exists on disk
+                        // remove from `in_progress` if it does
+                        if path.exists() {
+                            let mut sh = shared.lock().unwrap();
+                            sh.in_progress.remove(&path);
+                            continue;
+                        }
 
-                        flush_all(&mut queue, &pool_tx);
-                        break;
-                    }
-                    Ok(ChunkWriterMessage { sg, digest, chunk_type }) => {
-                        let path =
-                            self.chunk_path_by_digest(&digest, chunk_type);
-                        if !path.exists() &&
-                           !in_progress.lock().unwrap().contains(&path) {
+                        // write file to disk
+                        let tmp_path = path.with_extension("tmp");
+                        // Workaround
+                        // https://github.com/rust-lang/rust/issues/33707
+                        let _ = fs::create_dir_all(path.parent().unwrap());
 
+                        fs::create_dir_all(path.parent().unwrap()).unwrap();
+                        let mut chunk_file = fs::File::create(&tmp_path)
+                            .unwrap();
 
-                            if queue.len() == queue.capacity() {
-                                flush_one(&mut queue, &pool_tx);
-                            }
+                        let mut bytes_written = 0;
+                        for data_part in sg.iter() {
+                            chunk_file.write_all(&data_part).unwrap();
+                            bytes_written += data_part.len() as u64;
+                        }
 
-                            let tmp_path = path.with_extension("tmp");
-                            fs::create_dir_all(path.parent().unwrap()).unwrap();
-                            let mut chunk_file = fs::File::create(&tmp_path)
-                                .unwrap();
-                            for data_part in sg.iter() {
-                                chunk_file.write_all(&data_part).unwrap();
-                                write_stats.new_bytes += data_part.len() as u64;
-                            }
-                            write_stats.new_chunks += 1;
+                        flush_one(&shared, chunk_file, path);
```

Comment:
I would remove the flush_one function and just put the body of the function here. You acquire a lock in the function as well as a lock below this function call. It will keep the complexity of the function down and saves one lock operation.